erofs/erofs.go

```diff
@@ -0,0 +1,744 @@
+// This package is a small go "library" (read: exec wrapper) around the
+// mkfs.erofs binary that provides some useful primitives.
+package erofs
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+	"machinerun.io/atomfs/log"
+	"machinerun.io/atomfs/mount"
+)
+
+var checkZstdSupported sync.Once
+var zstdIsSuspported bool
+
+var exPolInfo struct {
+	once   sync.Once
+	err    error
+	policy *ExtractPolicy
+}
+
+// ExcludePaths represents a list of paths to exclude in a erofs listing.
+// Users should do something like filepath.Walk() over the whole filesystem,
+// calling AddExclude() or AddInclude() based on whether they want to include
+// or exclude a particular file. Note that if e.g. /usr is excluded, then
+// everyting underneath is also implicitly excluded. The
+// AddExclude()/AddInclude() methods do the math to figure out what is the
+// correct set of things to exclude or include based on what paths have been
+// previously included or excluded.
+type ExcludePaths struct {
+	exclude map[string]bool
+	include []string
+}
+
+type erofsFuseInfoStruct struct {
+	Path           string
+	Version        string
+	SupportsNotfiy bool
+}
+
+var once sync.Once
+var erofsFuseInfo = erofsFuseInfoStruct{"", "", false}
+
+func NewExcludePaths() *ExcludePaths {
+	return &ExcludePaths{
+		exclude: map[string]bool{},
+		include: []string{},
+	}
+}
+
+func (eps *ExcludePaths) AddExclude(p string) {
+	for _, inc := range eps.include {
+		// If /usr/bin/ls has changed but /usr hasn't, we don't want to list
+		// /usr in the include paths any more, so let's be sure to only
+		// add things which aren't prefixes.
+		if strings.HasPrefix(inc, p) {
```

Comment:
But what if both /usr/bin/cal and /usr/bin/calendar have changed?  Will this check end up possibly ignoring /usr/bin/cal?